hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/opt/logical/SlidingWindowFilterTransposeLogicalRule.java

```diff
@@ -22,64 +22,82 @@
- * window.
+ * A {@link Calc} condition reading from a {@link SlidingWindow}
+ * will be moved before the sliding window as {@link Filter}
  */
 @Value.Enclosing
-public class SlidingWindowFilterTransposeRule extends RelRule<Config> implements TransformationRule {
+public class SlidingWindowFilterTransposeLogicalRule extends RelRule<Config> implements TransformationRule {
 
     @Value.Immutable
     public interface Config extends RelRule.Config {
-        Config DEFAULT = ImmutableSlidingWindowFilterTransposeRule.Config.builder()
-                .description(SlidingWindowFilterTransposeRule.class.getSimpleName())
+        Config DEFAULT = ImmutableSlidingWindowFilterTransposeLogicalRule.Config.builder()
+                .description(SlidingWindowFilterTransposeLogicalRule.class.getSimpleName())
                 .operandSupplier(b0 -> b0
-                        .operand(Filter.class)
+                        .operand(CalcLogicalRel.class)
+                        .predicate(calc -> calc.getProgram().getCondition() != null)
                         .trait(LOGICAL)
                         .inputs(b1 -> b1
                                 .operand(SlidingWindow.class).anyInputs()))
                 .build();
 
         @Override
         default RelOptRule toRule() {
-            return new SlidingWindowFilterTransposeRule(this);
+            return new SlidingWindowFilterTransposeLogicalRule(this);
         }
     }
 
-    public static final RelOptRule STREAMING_FILTER_TRANSPOSE = new SlidingWindowFilterTransposeRule(Config.DEFAULT);
+    public static final RelOptRule STREAMING_FILTER_TRANSPOSE = new SlidingWindowFilterTransposeLogicalRule(DEFAULT);
 
-    protected SlidingWindowFilterTransposeRule(Config config) {
+    protected SlidingWindowFilterTransposeLogicalRule(Config config) {
         super(config);
     }
 
     @Override
     public void onMatch(RelOptRuleCall call) {
-        final Filter filter = call.rel(0);
+        final CalcLogicalRel calc = call.rel(0);
         final SlidingWindow sw = call.rel(1);
 
+        RexProgram program = calc.getProgram();
         RexVisitorImpl<Void> visitor = new RexVisitorImpl<Void>(true) {
             @Override
             public Void visitInputRef(RexInputRef ref) {
                 int index = ref.getIndex();
                 if (index == sw.windowStartIndex() || index == sw.windowEndIndex()) {
+                    // TODO[sasha]: convert into ShouldnotExecuteRel (after #20996 merge)
                     throw QueryException.error("Can't apply filter criteria to window bounds");
                 }
                 return super.visitInputRef(ref);
             }
         };
-        filter.getCondition().accept(visitor);
 
-        final Filter newFilter = filter.copy(filter.getTraitSet(), sw.getInput(), filter.getCondition());
+        program.expandLocalRef(program.getCondition()).accept(visitor);
+
+        final FilterLogicalRel newFilter = new FilterLogicalRel(
```

Comment:
Shouldn't we create Calc here with this filter, and an identity projection? Then we can maybe eliminate all `Filter*Rel` and `Filter*Rule`.